cabal-install/bootstrap.py

```diff
@@ -0,0 +1,436 @@
+
+def package_cabal_url(package: PackageName, version: Version, revision: int) -> str:
+    return f'https://hackage.haskell.org/package/{package}-{version}/revision/{revision}.cabal'
+
+def verify_sha256(expected_hash: SHA256Hash, f: Path):
+    h = hash_file(hashlib.sha256(), f.open('rb'))
+    if h != expected_hash:
+        raise BadTarball(f, expected_hash, h)
+
+def fetch_package(package: PackageName,
+                  version: Version,
+                  src_sha256: SHA256Hash,
+                  revision: Optional[int],
+                  cabal_sha256: Optional[SHA256Hash],
+                  ) -> Path:
+    import urllib.request
+
+    # Download source distribution
+    out = PACKAGES / (f'{package}-{version}.tar.gz')
+    if not out.exists():
+        print(f'Fetching {package}-{version}...')
+        out.parent.mkdir(parents=True, exist_ok=True)
+        url = package_url(package, version)
+        with urllib.request.urlopen(url) as resp:
+            shutil.copyfileobj(resp, out.open('wb'))
+
+    # Download revised cabal file
+    cabal_file = PACKAGES / f'{package}.cabal'
+    if revision is not None and not cabal_file.exists():
+        assert cabal_sha256 is not None
+        url = package_cabal_url(package, version, revision)
+        with urllib.request.urlopen(url) as resp:
+            shutil.copyfileobj(resp, cabal_file.open('wb'))
+            verify_sha256(cabal_sha256, cabal_file)
+
+    verify_sha256(src_sha256, out)
+    return out
+
+def read_bootstrap_deps(path: Path) -> List[BootstrapDep]:
+    obj = json.load(path.open())
+
+    def from_json(o: dict) -> BootstrapDep:
+        o['source'] = PackageSource(o['source'])
+        return BootstrapDep(**o)
+
+    return [from_json(dep) for dep in obj['dependencies']]
+
+def write_bootstrap_deps(path: Path, deps: List[BootstrapDep]):
+    def to_json(dep: BootstrapDep) -> object:
+        return {
+            'package': dep.package,
+            'version': dep.version,
+            'source': dep.source.value,
+            'src_sha256': dep.src_sha256,
+            'revision': dep.revision,
+            'cabal_sha256': dep.cabal_sha256,
+        }
+
+    obj = {
+        'dependencies': [to_json(dep) for dep in deps],
+    }
+    json.dump(obj, path.open('w'), indent=2)
+
+def install_dep(dep: BootstrapDep, ghc: Compiler) -> None:
+    if dep.source == PackageSource.PREEXISTING:
+        # We expect it to be in the compiler's bootstrap package set
+        subprocess.run([str(ghc.ghc_pkg_path), 'describe', f'{dep.package}-{dep.version}'],
+                       check=True, stdout=subprocess.DEVNULL)
+        print(f'Using {dep.package}-{dep.version} from GHC...')
+        return
+
+    elif dep.source == PackageSource.HACKAGE:
+        assert dep.src_sha256 is not None
+        tarball = fetch_package(dep.package, dep.version, dep.src_sha256,
+                                dep.revision, dep.cabal_sha256)
+        subprocess.run(['tar', 'xf', tarball.resolve()],
```

Comment:
Please use `zxf` to support non-GNU tar versions.

I confirmed that I can this patched script with ghc 8.10.1 on OpenBSD.

Reply:
Fair point.